File: ocaml/lib/some_parsers.ml
open Maybe
open Parser

let anyChar = P (fun inp -> if String.length inp > 0
    then just (inp.[0], String.sub inp 1 ((String.length inp) - 1))
    else nothing)

let satisfy pred = anyChar >>= (fun x -> if pred x then pure x else empty)

let char c = satisfy (fun x -> x == c)

let empty_string = pure String.empty

let rec some p = (fun c -> fun s -> String.cat (String.make 1 c) s) <$> p <*> (fun () -> many p)
and many p = (some p) <||> empty_string

let spaces = many (char ' ')

let between_get open_ close_ fp = (open_ >> fp) >>= (fun x -> close_ >> (fun () -> pure x))

let between open_ close_ p = between_get open_ close_ (fun () -> p)

let token p = between spaces spaces p

let symbol c = token (char c)

let is_alpha alpha = 
    match alpha with 
      'a' .. 'z' -> true
    | 'A' .. 'Z' -> true 
    | _ -> false;;

let is_digit digit =
    match digit with
     '0' .. '9' -> true
    | _ -> false;; 

let alnum = satisfy (fun c -> is_alpha c || is_digit c || c == '_')

let name n = token (some alnum) >>= (fun s -> if String.equal s n then pure n else empty)

let optional_s p = p <||> empty_string

let optional_c p = optional_s ((fun c -> String.make 1 c) <$> p)

let sign = optional_c (char '+' <||> char '-')

let usign = optional_c (symbol '+' <||> symbol '-')

let digits = many (satisfy is_digit)

let double = token(sign >>=
    (fun sign_part -> digits >>=
    (fun int_part  -> optional_s (char '.' >>> digits) >>=
    (fun frac_part -> optional_s (((char 'e' <||> char 'E') >>> sign) >>=
        (fun exp_sign -> some (satisfy is_digit) >>=
        (fun exp_digits -> pure (String.cat exp_sign exp_digits)))) >>=
    (fun exp_part  -> if String.length int_part > 0 || String.length frac_part > 0 then
        pure (Float.of_string (sign_part ^ int_part ^
            (if String.length frac_part > 0 then "." ^ frac_part else "") ^
            (if String.length exp_part > 0 then "e" ^ exp_part else "")))
        else empty)))))

let rest p ff op x = (op >>= (fun f -> p >>= (fun y -> ff (f x y)))) <|> (fun () -> pure x)

let rec rest_l p op x = rest p (fun y -> rest_l p op y) op x

let chainl1 p op negate_first = p >>= (fun x -> rest_l p op (if negate_first then -.x else x))

let rec rest_r p op x = rest (chainr1 p op) pure op x
and chainr1 p op = p >>= (fun x -> rest_r p op x)

File: R/src/some_parsers.R
source("../src/parser.R")

empty_string <- pure("")

some <- function(p) {
  p %>% parser_map(function(c) { function(s) { paste0(c, s) }}) %>% parser_apply( function(){ many(p) } )
}

many <- function(p) {
  some(p) %>% parser_or_else(function() { empty_string })
}

any_char <- as_parser(function(inp) {
  if(nchar(inp) > 0)
    just(list(substring(inp, 1, 1), substring(inp, 2)))
  else nothing()
})

satisfy <- function(pred) {
  any_char %>% parser_and_then(function(c) { if(pred(c)) pure(c) else empty })
}

char <- function(c) {
  satisfy(function(x){ x == c })
}

spaces <- many(satisfy(function(c){ grepl("\\s", c) }))
  
between <- function(open, close, fp){
  open %>% parser_skip(function(){ fp() %>% parser_and_then(function(x){ close %>% parser_skip(function(){ pure(x) }) }) })
}

token <- function(p) {
  between(spaces, spaces, function(){ p })
}

symbol <- function(c) {
  char(c) %>% token()
}

alnum <- satisfy(function(c){ grepl("[[:alnum:]]", c) || c == '_' })

name <- function(n) {
  some(alnum) %>% parser_and_then(function(n2){ if(n2 == n) pure(n) else empty }) %>% token()
}

optional_s <- function(p) {
  p %>% parser_or_else(function(){ empty_string })
}

digit = satisfy( function(c){ grepl("\\d", c) })

digits <- many(digit)

sign <- optional_s(char('+') %>% parser_or_else(function(){ char('-') }))

usign <- optional_s(symbol('+') %>% parser_or_else(function(){ symbol('-') }))

double <- token(digits %>%
  parser_and_then(function(int_part) { optional_s(char('.') %>% parser_skip(function() { digits })) %>%
  parser_and_then(function(frac_part) { optional_s(char('e') %>% parser_or_else(function() { char('E') }) %>% parser_skip(function(){ sign }) %>%
    parser_and_then(function(exp_sign) { some(digit) %>%
    parser_and_then(function(exp_digits) { pure(paste0(exp_sign, exp_digits)) }) })) %>%
  parser_and_then(function(exp_part) { if(nchar(int_part) > 0 || nchar(frac_part) > 0)
   pure(as.numeric(paste0(int_part,
     if(nchar(frac_part) > 0) paste0('.', frac_part) else "",
     if(nchar(exp_part) > 0) paste0('e', exp_part) else "")))
   else empty }) }) }))

rest <- function(fp, ff, op, x) {
  op %>% parser_and_then(function(f) {
    fp() %>% parser_and_then(function(y) { ff(f(x, y)) })
  }) %>% parser_or_else(function(){ pure(x) })
}

rest_l <- function(p, op, x) {
  rest(function(){ p }, function(y){ rest_l(p, op, y) }, op, x)
}

chainl1 <- function(p, op, negate_first) {
  p %>% parser_and_then(function(x) { rest_l(p, op, if(negate_first) -x else x) })
}

rest_r <- function(p, op, x) {
  rest(function(){ chainr1(p, op) }, function(y){ pure(y) }, op, x)
}

chainr1 <- function(p, op) {
  p %>% parser_and_then(function(x) { rest_r(p, op, x) })
}
